pennylane/capture/base_interpreter.py

```diff
@@ -0,0 +1,471 @@
+    in (h,) }
+
+    Accumulation transforms, like device execution or conversion to tapes, may need to flatten out
+    the higher order primitive to execute it.
+
+    .. code-block:: python
+
+        class AccumulateOps(PlxprInterpreter):
+
+            def __init__(self, ops=None):
+                self.ops = ops
+
+            def setup(self):
+                if self.ops is None:
+                    self.ops = []
+
+            def interpret_operation(self, op):
+                self.ops.append(op)
+
+        @AccumulateOps.register_primitive(qml.capture.primitives.for_loop_prim)
+        def _(self, *invals, jaxpr_body_fn, n_consts):
+            start, stop, step = invals[0], invals[1], invals[2]
+            consts = invals[3 : 3 + n_consts]
+            state = invals[3 + n_consts :]
+
+            for i in range(start, stop, step):
+                state = copy(self).eval(jaxpr_body_fn, consts, i, *state)
+            return state
+
+    >>> @qml.for_loop(3)
+    ... def loop(i, x):
+    ...     qml.RX(x, i)
+    ...     return x
+    >>> accumulator = AccumlateOps()
+    >>> accumulator(loop)(0.5)
+    >>> accumulator.ops
+    [RX(0.5, wires=[0]), RX(0.5, wires=[1]), RX(0.5, wires=[2])]
+
+    In this case, we need to actually evaluate the jaxpr 3 times using our interpreter. If jax's
+    evalutation interpreter ran it three times, we wouldn't actually manage to accumulate the operations.
+
+
+    """
+
+    _env: dict
+    _primitive_registrations: dict["jax.core.Primitive", Callable] = {}
+
+    def __init_subclass__(cls) -> None:
+        cls._primitive_registrations = copy(cls._primitive_registrations)
+
+    def __init__(self):
+        self._env = {}
+
+    @classmethod
+    def register_primitive(cls, primitive: "jax.core.Primitive") -> Callable[[Callable], Callable]:
+        """Registers a custom method for handling a primitive
+
+        Args:
+            primitive (jax.core.Primitive): the primitive we want custom behavior for
+
+        Returns:
+            Callable: a decorator for adding a function to the custom registrations map
+
+        Side Effect:
+            Calling the returned decorator with a function will place the function into the
+            primitive registrations map.
+
+        .. code-block:: python
+
+            my_primitive = jax.core.Primitive("my_primitve")
+
+            @Interpreter_Type.register(my_primitive)
+            def handle_my_primitive(self: Interpreter_Type, *invals, **params)
+                return invals[0] + invals[1] # some sort of custom handling
+
+        """
+
+        def decorator(f: Callable) -> Callable:
+            cls._primitive_registrations[primitive] = f
+            return f
+
+        return decorator
+
+    # pylint: disable=unidiomatic-typecheck
+    def read(self, var):
+        """Extract the value corresponding to a variable."""
+        return var.val if type(var) is jax.core.Literal else self._env[var]
```

Comment:
Why not use `isinstance` here?

Reply:
We probably could.  This was the syntax used in the tutorial on writing an interpreter, but it does cause pylint to complain.

Reply:
Probably best to use `isinstance`. Then, the pylint warning suppression can be removed as well.